### PR TITLE
COMP: Add completion for path-based keys in Cargo.toml

### PR DESCRIPTION
**Pull request by mibac138**
path-completion

### CODE REVIEW

**Comment by vlad20012** [comment]
Hi!
Take a look it the implementation of include macro path reference. 
Peek 2019-06-06 21-35

See RsLitExprReferenceContributor/RsFileReferenceProvider/RsLiteralFileReferenceSet. It is more idiomatic and also provides go-to-declaration, rename, find usages (and really should provide error reporting if file is undefined, but it is not work for include macro for some reason), etc. I want to implement it this way a lot time ago, but it's rather hard (it lest, if make it correctly).

The most complex part is that strings can have [escape codes](https://github.com/toml-lang/toml#string) and we should be able to both (!) escape and unescape it.
A brief explanation of how to make it work (mostly I quote how it is implemented for Rust):
1. Implement a lexer for TOML escape sequences (see [TOML escape codes](https://github.com/toml-lang/toml#string) ). Rust analogue - RustEscapesLexer. It can be used directly for highlighting (see RsHighlightingLexer). Also see tests RsEscapesLexingTestCase
2. Implement string escape/unescape. See RsEscapeUtils.kt. Also see tests RsEscapesUtilsTest.kt. This can also be used in spell checking. See RsSpellcheckingStrategy.
3. Implement LiteralTextEscaper. See RsNormalStringLiteralEscaper. This is not useful itself, so should be done in a couple with the next options.
4. Implement PsiLanguageInjectionHost interface for TomlLiteral. There should be used LiteralTextEscaper from the prev option and also updateText should be implemented with text escaping. See RsLitExprMixin (but as I see updateText implemented incorrectly there). At this point it's possible to provide language injections to TOML string literals (see RsRegExpInjector). But I don't think it's useful
5. Implement ElementManipulator for TOML literals. See RsStringLiteralManipulator (again, as I see it's implemented incorrectly b/c is doesn't escape the text). `getRangeInElement` method is the most necessary. Finally.
6. Also it can be useful (for all the above) to make something like RsLiteralKind for TOML. (offsets are the most useful)

Note that ElementManipulator and PsiLanguageInjectionHost (and so LiteralTextEscaper) implementation are needed to make RsFileReferenceProvider/FileReferenceSet work.

Please let me know it you want to work on it.
Also let me know it you want this PR to be merged (I'll just leave few comments). It's not ideal, but it's better than nothing. And thank you!

File: toml/src/main/kotlin/org/rust/toml/CargoTomlPathCompletion.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            File -> if (file.isDirectory) DEFAULT_PRIORITY else 1.0
+            Directory -> if (file.isDirectory) 1.0 else DEFAULT_PRIORITY
+            Any -> DEFAULT_PRIORITY
+        }
+    }
+
+    override fun addCompletions(parameters: CompletionParameters, context: ProcessingContext, result: CompletionResultSet) {
+        val keyValue = getClosestKeyValueAncestor(parameters.position) ?: return
+        val containingFile = keyValue.containingFile.originalFile.virtualFile ?: return
+        val path = keyValue.value?.text
```
Comment (vlad20012) [inline toml/src/main/kotlin/org/rust/toml/CargoTomlPathCompletion.kt]:
```suggestion
        val path = CompletionUtil.getOriginalElement(keyValue.value)?.text
```

After this change you will be able to remove `?.substringBefore("IntellijIdeaRulezzz")` :)

Comment (mibac138) [reply toml/src/main/kotlin/org/rust/toml/CargoTomlPathCompletion.kt]: Thank you

**Comment by mibac138** [comment]
@vlad20012 After some consideration I'd prefer to have this PR merged and then do a follow-up PR with the changes you described. This way there will be at least some completion in paths and when I'll have more time and once I implement what you described I'll submit another PR enhancing this one.

**Comment by Undin** [comment]
> After some consideration I'd prefer to have this PR merge

@mibac138 sorry but I don't think it's a good idea
Temporary things tend to become permanent.
It will take some of our time to review this PR. Including the fact that the current implementation is not a subset of the desired one and should be replaced in future, I think it's better not to merge this version at all

**Comment by mibac138** [comment]
@Undin sure, I totally get that. I'll open a new PR once I implement this as `@vlad20012` described.

**Comment by vlad20012** [comment]
@mibac138 It may not be useless to write somewhere ([gitter](https://gitter.im/intellij-rust/intellij-rust or open an issue)) about what you're going to work (if it's not trivial) to prevent such situations...

**Pull request closed without merging**